[wgsl-in] Implement template list discovery and handle all type resolution in the lowerer

teoxoy opened the pull request:
**Connections**
Resolves https://github.com/gfx-rs/wgpu/issues/4501.
Resolves https://github.com/gfx-rs/wgpu/issues/4406.
Resolves https://github.com/gfx-rs/wgpu/issues/4405.

**Description**
Almost all commits prior to `implement template list discovery` are refactors to get the frontend to look more similar to the spec and to make the needed changes easier to implement (most notably making the `Tokenizer` no longer impl `Clone`).

The template list discovery implementation differs from the spec only in structure (not in behavior) since we can do it lazily after tokenization.

The last 3 substantial commits move all type resolution to the lowerer since all identifiers that are part of types need to be resolved the same as all other identifiers.

**Testing**
Changed existing tests and added new ones.

**Squash or Rebase?**
Rebase.

teoxoy committed:
[wgsl] update reserved list to latest CR

teoxoy committed:
add `optionally_typed_ident` fn

teoxoy committed:
[wgsl] rename `type_decl` to `type_specifier`

teoxoy committed:
rename `assignment_statement` to `variable_updating_statement`

teoxoy committed:
rename `general_expression` to `expression`

teoxoy committed:
remove `next_scalar_generic`

teoxoy committed:
remove `peek_ident_with_span`

teoxoy committed:
update `lhs_expression` to be in line with the spec

teoxoy committed:
rename `postfix` to `component_or_swizzle_specifier`

teoxoy committed:
remove `lexer.clone()`

teoxoy committed:
rename `function_statement` to `func_call_statement`

teoxoy committed:
rename `function_call_or_assignment_statement` to `func_call_or_variable_updating_statement`

teoxoy committed:
use `lexer.next()` instead of `peek()` for most statements

teoxoy committed:
flatten match

teoxoy committed:
add function that parses variable, value, func call and variable updating statements

teoxoy committed:
use `lexer.skip()` in one more place

teoxoy committed:
rename `Lexer` `skip` to `next_if`

teoxoy committed:
use `lexer.next()` in `primary_expression`

teoxoy committed:
rename `function_call` to `call_expression`

teoxoy committed:
use `lexer.next()` in `unary_expression`

teoxoy committed:
avoid recursing into `unary_expression`

teoxoy committed:
remove `Clone` derive from the `Lexer`

also fixes spans in some cases

teoxoy committed:
implement template list discovery

teoxoy committed:
add `Lexer::span_with_start`

teoxoy committed:
add `TemplateArgsStart` & `TemplateArgsEnd` tokens

teoxoy committed:
handle `var` decl template list in the lowerer

teoxoy committed:
handle `bitcast`, `coopLoad` and construction exprs in the lowerer

teoxoy committed:
handle all type resolution in the lowerer

teoxoy committed:
add `shadowing_predeclared_types` test

teoxoy committed:
add `test_template_list` test

teoxoy committed:
document `consume_tokens`

teoxoy committed:
adjust GLSL snapshots

jimblandy commented:
I can take review here - I've been looking into this for a long time.

jimblandy commented:
@teoxoy It seems like a bunch of these early commits would be fine as independent cleanup PRs, like 42338fff3 and 4adede577.

jimblandy commented on naga/src/front/wgsl/error.rs:
```diff
@@ -480,13 +488,16 @@ impl<'a> Error<'a> {
                         Token::End => "end".to_string(),
                     },
                     ExpectedToken::Identifier => "identifier".to_string(),
+                    ExpectedToken::LhsExpression => "lhs_expression".to_string(),
```
I think `lhs_expression` is spec jargon. Isn't there something we could say here that would be more readily recognized by WGSL programmers? "expression to be assigned to"? "destination for assignment"?

(Spec jargon is fine when we need to be precise and there's nothing better.)

teoxoy replied:
"destination for assignment"/"assignment destination" sound good.
It's also used by increment/decrement statements but it's probably ok?

teoxoy replied:
I ended up changing this as part of the last commit to:

```
LHS expression (identifier component_or_swizzle_specifier?, (`lhs_expression`) component_or_swizzle_specifier?, &`lhs_expression`, *`lhs_expression`)
```

It's a lot more verbose but I couldn't figure out what would be better since it's now really only used in cases where an LHS expression is expected.

jimblandy commented:
So far I've reviewed the following commits and they all look fine:

```
527cf31a1 | * rename `function_call_or_assignment_statement` to `func_call_or_variable_updating_statement`
16c2777ba | * rename `function_statement` to `func_call_statement`
086c0e4c3 | * remove `lexer.clone()`
fe63d0753 | * rename `postfix` to `component_or_swizzle_specifier`
47108f674 | * update `lhs_expression` to be in line with the spec
75fbca1ed | * remove `peek_ident_with_span`
ed6a03205 | * remove `next_scalar_generic`
e99eb0149 | * rename `general_expression` to `expression`
54fe3ca2e | * rename `assignment_statement` to `variable_updating_statement`
4adede577 | * [wgsl] rename `type_decl` to `type_specifier`
42338fff3 | * add `optionally_typed_ident` fn
```

teoxoy committed:
[wgsl-out] avoid shadowing built-ins

teoxoy commented:
I removed the last commit that was adding a CTS test since it wasn't passing due to unrelated reasons and pushed 2 more commits to fix the issues CI surfaced.

teoxoy commented:
> @teoxoy It seems like a bunch of these early commits would be fine as independent cleanup PRs, like [42338ff](https://github.com/gfx-rs/wgpu/commit/42338fff3a277735ae81e57fc01d1051da3dc02c) and [4adede5](https://github.com/gfx-rs/wgpu/commit/4adede5773484471dbbcf1f302713b05b123ae51).

They could but that would be lots of separate PRs to file and orchestrate since I think some of them depend on others. I could open a PR just with the refactor commits but then this PR would need to depend on that new one. Let me know.

jimblandy commented on naga/src/front/wgsl/parse/mod.rs:
```diff
@@ -2116,391 +1216,285 @@ impl Parser {
     ) -> Result<'a, ()> {
         self.track_recursion(|this| {
             this.push_rule_span(Rule::Statement, lexer);
+
             match lexer.peek() {
```
It would be nice to have a comment explaining why this `match` statement shouldn't be folded into the following `match`.

I think it's because `Parser::block` assumes the lexer is sitting before the block, but the following `match` consumes the block's first token, and while `Parser::block` could certainly accept the first token as an argument, we'd have to change a lot of callers. Is that right?

teoxoy replied:
Yes that's right. The reason I haven't changed `Parser::block` to take the first token (always a `{`) is because there is usually a single path to a block start in the grammar; `statement` is the only exception where there are multiple choices.

teoxoy replied:
I added a comment.

jimblandy commented on naga/src/front/wgsl/parse/mod.rs:
```diff
@@ -2005,106 +1055,156 @@ impl Parser {
+    }
+
+    /// Parses variable_or_value_statement, func_call_statement and variable_updating_statement
+    fn variable_or_value_or_func_call_or_variable_updating_statement<'a>(
```
This deserves a comment mentioning that this is equivalent to the `for_init` production in the WGSL spec, but that it's also used for parsing these forms when they appear within a block, hence the longer name.

jimblandy commented on naga/tests/naga/wgsl_errors.rs:
```diff
@@ -396,11 +396,11 @@ fn bad_for_initializer() {
             }
         "#,
-        r#"error: for(;;) initializer is not an assignment or a function call: `{}`
+        r#"error: expected lhs_expression, found "{"
```
This is a less clear error message than the original. Overall there are a lot of cases like this. It seems like we end up in a deep call stack before we finally give up parsing:

- `variable_or_value_or_func_call_or_variable_updating_statement`
- `func_call_or_variable_updating_statement`
- `variable_updating_statement`
- `lhs_expression`

And then it's `lhs_expression` that generates the error. This error complaining about an LHS expression is then propagated all the way out into these functions which actually were ready for much more general things, so it's kind of misleading.

One fix would be to have the callers detect the too-specific `ExpectedToken` values and replace them with something more general, like a new `ForInit` variant.

Another fix would be to have this family of functions accept `Option<ExpectedToken>` arguments that say what to complain about if parsing fails. That way `variable_or_value_or_func_call_or_variable_updating_statement` could be passed something about `for` initializers when parsing "for" statements, but `None` when called just to parse the statement.

It's probably okay to handle this in a follow-up PR, but it is a misleading regression, so we shouldn't let it languish for very long.

teoxoy replied:
This is a good point, I added another commit that resolves this.

jimblandy commented on naga/src/front/wgsl/parse/lexer.rs:
```diff
@@ -35,24 +36,118 @@ fn consume_any(input: &str, what: impl Fn(char) -> bool) -> (&str, &str) {
+        }
+    }
+
+    loop {
```
Won't this enter an infinite loop if `consume_token` returns `Token::End` while `pending` is not empty?

teoxoy replied:
It will, good catch, I keep forgetting `Token::End` is a token type 😅.

teoxoy replied:
I added a test and pushed the fix.

jimblandy reviewed (changes requested):
All the refactoring changes up to `4ebbc67df * implement template list discovery` look good. I think there's a bug in the TLD code itself.

andyleiserson commented:
I pushed a version of this that is rebased on top of #7339 to https://github.com/andyleiserson/wgpu/commits/tld-rebased. The last commit resolves a semantic merge conflict with #7339. There were also merge conflicts with trunk in 314ba9bceba2162b6a545bc45d6665b66fba0dc4 and cf9e1113923647d3fad54d11783266f7914c3f3c that I resolved, and there is a merge conflict with #8545 that I was less sure how to handle (that change adds enable extension checks that need to be relocated somewhere else following the template list discovery changes).

teoxoy commented:
@andyleiserson thanks for tracking the changes! I pushed a fully rebased version.

teoxoy committed:
[wgsl] pass expected token to functions that might not match any tokens

This improves errors making them more general where appropriate.

jimblandy committed:
[naga wgsl-in] Rename `component_or_swizzle_specifier` arg.

In `naga::front::wgsl::parse::Parser::component_or_swizzle_specifier`,
since `span_start` sounds like the start of a span, rename that
argument to `expr_start`, since it is the span of the start of the
expression that ends with the component or swizzle specifier.

jimblandy committed:
[naga wgsl-in] Doc fixes.

- Clarify that various `foo_statement` functions on `Parser` do not
  consume the statement's final `;`.

- Various other doc fixes throughout.

jimblandy commented:
I pushed two very minor tweaks from today's reviewing.

teoxoy commented:
The first one looks good but the 2nd one assumes that those statements always end with a semicolon. Off the top of my head they are also used in the for loop header.

jimblandy committed:
[naga wgsl-in] Use `Option::is_some_and` where appropriate.

jimblandy committed:
[naga wgsl-in] Rename `consume_tokens` to `discover_template_lists`.

jimblandy commented:
I pushed two very minor tweaks from today's reviewing.

> The first one looks good but the 2nd one assumes that those statements always end with a semicolon. Off the top of my head they are also used in the for loop header.

Okay, right - I fixed these.

Pushed a few more, and some tweaks, please take a look.

jimblandy commented:
The template list discovery commit looks good to me.

teoxoy commented on naga/src/front/wgsl/parse/lexer.rs:
```diff
@@ -129,6 +223,9 @@ fn discover_template_lists<'a>(
 
+        // The WGSL spec's template list discovery algorithm processes
+        // the entire source at once, but Naga would rather limit its
+        // lookahead to the actual text that needs to be
```
Missing text?

jimblandy replied:
Yes, thanks.

jimblandy committed:
[naga wgsl-in] Remove unnecessary dependencies.

Avoid adding entries to `ExpressionContext::unresolved` for
identifiers that `ExpressionContext::ident_expr` has already
processed.

At the moment, there is no way these could cause a bug: they only
occur after we've recognized a call/construction expression, and those
can never refer to function-local definitions, so the string in
question is certain to have already been added to `unresolved` by
`ExpressionContext::ident_expr`. But they're redundant, and if Naga
ever supported nested functions or local type aliases, these could end
up adding extraneous dependencies.

jimblandy commented:
Okay, some more docs, but also some actual code changes. I think those dependency insertions are redundant; see the commit message for details. It's conceptually nice if we always pair creation of `IdentExpr::Unresolved` values with insertions into `ExpressionContext::unresolved`.

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2562,6 +2544,231 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
         }
     }
 
+    fn enumerant(
```
Would this function and `var_address_space` make more sense as methods on `ExpressionContext`?

teoxoy replied:
Oh yeah - I can update those.

jimblandy committed:
[naga wgsl-in] Rename `TemplateListIter::scalar` to `scalar_ty`.

The name `scalar` suggests that it returns a scalar value, and
`scalar_ty` is analogous to `ty`.

teoxoy commented:
Changes look good, thanks!

teoxoy committed:
[wgsl-in] move `enumerant` & `var_address_space` functions on `ExpressionContext`

jimblandy committed:
[naga wgsl-in] Use typical abbreviations in `map_predeclared_type`.

jimblandy committed:
[naga wgsl-in] Minor rearrangement to flatten control flow.

Use `if let`, `let else`, and `Option` methods to reduce nesting and
clarify success paths.

jimblandy committed:
[naga wgsl-in] Doc fixes.

jimblandy commented on naga/src/proc/namer.rs:
```diff
@@ -90,6 +90,7 @@ pub struct Namer {
     /// The last numeric suffix used for each base name. Zero means "no suffix".
     unique: FastHashMap<String, u32>,
     keywords: &'static KeywordSet,
+    builtin_identifiers: &'static KeywordSet,
```
Given that WGSL provides no way for code injected by Naga to refer to predeclared objects if they are shadowed (as one could do in Rust, say, by writing `std::` or `core::`, or in C++ with a leading `::`), it seems we have no choice but to tell the namer to avoid letting user code use any of these names.

Would it be better to call this `predeclared_identifiers`, since "predeclared" is the term used by the WGSL specification?

teoxoy replied:
I think I avoided including "predeclared" in the name since more than half of them are built-in functions but those can be shadowed so maybe they should be called predeclared functions in the spec? We have too many names for things 😅.

jimblandy committed:
[naga] Simplify `proc::Namer`'s handling of struct member names.

Rather than passing `call_impl` an `ignore_builtin_identifiers` flag
and plumbing that through from its points of use, extend
`Namer::namespace` to suspend the `builtin_identifiers` table, as it
does for `unique` already. The resulting behavior is the same.

jimblandy commented:
In [7f1f953](https://github.com/gfx-rs/wgpu/pull/8386/commits/7f1f9534bd8bd0ebe8744cde70667dfceafb619b) I tried out a different approach to handle struct members. I think this is a bit cleaner; please let me know what you think.

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2748,8 +2978,72 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
             }
             None => {
+                match function_name {
+                    "bitcast" => {
```
Eventually, it would be nice to fold this in with all the other builtin functions (move the `tl.finish()` after the whole kit and kaboodle), because handling it early here makes one wonder if there's something special about `bitcast` other than its taking template parameters.

teoxoy replied:
The intent was to keep all the identifiers that expect template parameters up top to not forget to call `tl.finish()`. The problem is with early returns, if there are any, `tl.finish()` won't get to run. Maybe we can have some sort of builder that panics in its `Drop` impl if `finish()` is not called.

jimblandy replied:
Yeah, okay. I think the current approach is probably the cleanest and simplest.

jimblandy replied:
Idle thought: perhaps if we refactor `Lowerer::call`, there will be a single return path where `tl.finish()` could be called.

jimblandy commented:
Pre-existing, should not be fixed here, but: wow, `Lowerer::call` is waaay too long.

teoxoy commented:
> In [7f1f953](https://github.com/gfx-rs/wgpu/pull/8386/commits/7f1f9534bd8bd0ebe8744cde70667dfceafb619b) I tried out a different approach to handle struct members. I think this is a bit cleaner; please let me know what you think.

It is more appropriate, thanks!

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2562,6 +2583,202 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+        ctx: &mut ExpressionContext<'source, '_, '_>,
+        ty: conv::PredeclaredType,
+        tl: &mut TemplateListIter<'_, 'source>,
+        alias_name: Option<String>,
```
It is a little weird that `alias_name` is ignored for `RayDesc` and `RayIntersection`. But this seems to be a pre-existing problem.

jimblandy committed:
[naga wgsl-in] Don't permit template parameters after enumerants.

jimblandy committed:
[naga wgsl-in] Flatten control flow for clarity.

Adjust some methods of `TemplateListIter` to ensure that the typical
complete control flow through the function follows the outermost
nesting level.

jimblandy committed:
[naga wgsl-in]: Rename error variant field for clarity.

Rename `naga::front::wgsl::error::Error::MissingTemplateArg::arg` to
`description`, to make it a little more specific.

jimblandy reviewed (approved):
Okay, this looks great.

I fixed a minor problem, added a few comments, and did a rename. If those look okay, feel free to merge this.

jimblandy commented:
I'm working on rebasing this.

jimblandy commented:
I've rebased the initial refactoring commits onto `trunk`, addressing a few conflicts with other changes there, and pushed the partial rebase here: https://github.com/jimblandy/wgpu/tree/tld-refactors [edit: this is superseded, see next comment]

jimblandy commented:
Okay, I've got a bit more done, and made a new branch that I can push forward one commit at a time: https://github.com/jimblandy/wgpu/tree/tld-rebase-done

teoxoy commented:
> I fixed a minor problem, added a few comments, and did a rename. If those look okay, feel free to merge this.

They look good, thanks!

jimblandy commented:
I've pushed one more commit to `tld-rebase-done`, 158f0bbad05a201881c64819815ad4737625738c:

> handle `bitcast`, `coopLoad` and construction exprs in the lowerer

This got kind of involved, because the `coop_mat` types needed to be folded into the new system, and `coopLoad` takes a type parameter somewhat like `bitcast`.

jimblandy commented:
Okay, that's a full rebase of this PR on a recent wgpu trunk.

jimblandy commented:
@teoxoy Before we merge this, could you take a quick look at the way cooperative matrix types and functions are handled? I think just looking at the tip should be fine, no need to sort through all the commits.

jimblandy commented:
I just gave all the cooperative stuff a quick look, and I didn't notice anything that looked like it was doing things the old way.

- The only occurrences of the names of the `coop_mat` types are in `conv::map_predeclared_type`.
- Those map the types to a variant of `TypeGenerator`, like all our other builtin type generators.
- The only occurrences of the names of the coop functions are in `Lowerer::call`, where they belong.
- The only occurrences of `ir::TypeInner::CooperativeMatrix` in the WGSL front end are:
    - where we parse the types
    - where we check the types of the arguments to the coop functions
    - where we decline to perform automatic conversions on the inner scalars

jimblandy commented:
Since you marked this as "rebase", I consolidated my "doc fix" commits a bit, to keep the history cleaner.

jimblandy commented:
I have got to stop fiddling with this PR

teoxoy commented:
The cooperative matrix stuff looks good to me; as you mentioned it looks to me like we handle them the same way as the other predeclared types & builtin functions.

teoxoy commented:
Let's land this 🎉
